3578 order tags by frequency of use

bella-buchanan opened the pull request:
Resolves #3578 

**Problem**: Tags are presented to users in an unordered manner. It could help taggers to see tags in the order they are most frequently used. Additionally, it is important that mutually exclusive tags are presented next to each other for ease of understanding the tag. 

**Solution**: `TagTable.getTagsForCurrentCity` is an already implemented method that gets the tag frequencies for the city. In this PR, we use these counts to sort tags in `LabelController.getLabelTags`. 
##### Before/After screenshots

before (unsorted): 

<img width="398" alt="Image" src="https://github.com/user-attachments/assets/e71f027e-729c-4093-abeb-df21638f99f0" />

after (sorted and mutually exclusive pairing):

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/55f61209-6f14-4d4a-9b9d-07ecc053a0de" />

##### Testing instructions
1. 

##### Things to check before submitting the PR <!-- if something doesn't apply, just check the box or remove the line -->
<!-- You can check the box by replacing the space with an "x". So instead of "- [ ]" it would be "- [x]" -->
- [x] I've written a descriptive PR title. <!-- No need to include the issue number. Just a half sentence summary of the fix/feature! -->
- [x] I've added/updated comments for large or confusing blocks of code.
- [x] I've included before/after screenshots above.
- [ ] I've asked for and included translations for any user facing text that was added or modified.
- [ ] I've updated any logging. Clicks, keyboard presses, and other user interactions should be logged. If you're not sure how (or if you need to update the logging), ask Mikey. Then make sure the documentation on [this wiki page](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Descriptions-of-Logged-Events) is up to date for the logs you added/updated.
- [ ] I've tested on mobile (only needed for validation page).

bella-buchanan committed:
when querying tags, sort by label tag frequency

- adds a lookup to the  LabelTable to get global tag counts
- use counts to sort the tags order

bella-buchanan committed:
Revert "when querying tags, sort by label tag frequency"

This reverts commit c3051c2f43eea32f2973c6f8215afcc7b37e3d16.

bella-buchanan committed:
when querying tags, sort by label tag frequency - without the mess

- adds a lookup to the  LabelTable to get global tag counts
- use counts to sort the tags order

bella-buchanan committed:
ADD: group mutually exclusive tags together

misaugstad commented:
@bella-buchanan it looks like the mutually exclusive pairing isn't happening in all cases. One example below ("one button" and "two buttons" are mutually exclusive but not adjacent):
![Screenshot from 2025-03-03 14-06-42](https://github.com/user-attachments/assets/a4cadfd8-4729-4fe0-9659-f17f384b720c)

And just because they're sorted in that API doesn't mean that they're sorted in the actual UIs! I think that the ordering is being used by default on the Explore page, but not on the New Validate Beta page. Pasting something I wrote in the original ticket on that front:
>For the new Validate page, there is built-in functionality in the drop down search for adding tags that will sort based on a named field. I gave it a default name of "popularity" right now, and we would just need to add that property to the tag objects.

Here's an image from that UI:
![Screenshot from 2025-03-03 14-15-41](https://github.com/user-attachments/assets/16295471-9503-479e-baa1-7240e96992d4)

bella-buchanan committed:
Merge branch 'develop' into 3578-order-tags-by-frequency-of-use

bella-buchanan committed:
use popularity to sort tags

bella-buchanan committed:
sort by new field

bella-buchanan committed:
backend needs to give popularity based on mutually exclusive so they remain together

bella-buchanan committed:
sort tags by popularity on FE for explore page

bella-buchanan commented:
> @bella-buchanan it looks like the mutually exclusive pairing isn't happening in all cases. One example below ("one button" and "two buttons" are mutually exclusive but not adjacent): ![Screenshot from 2025-03-03 14-06-42](https://private-user-images.githubusercontent.com/6518824/418742531-a4cadfd8-4729-4fe0-9659-f17f384b720c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMzE2NzYsIm5iZiI6MTc0NDEzMTM3NiwicGF0aCI6Ii82NTE4ODI0LzQxODc0MjUzMS1hNGNhZGZkOC00NzI5LTRmZTAtOTY1OS1mMTdmMzg0YjcyMGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDhUMTY1NjE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDYxZDhkZTE4NmMzMjM2MDVlMWEyYmE1N2ZhMTk0NzFmMmFkMjcyMjRiYTk2NGJkMjQ2YzMzYTFlYjI0Y2M5NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NNv5xMcdUMIlUVvp0FxlTYvVtTWnenSNGD3goPj1XPc)
> 
> And just because they're sorted in that API doesn't mean that they're sorted in the actual UIs! I think that the ordering is being used by default on the Explore page, but not on the New Validate Beta page. Pasting something I wrote in the original ticket on that front:
> 
> > For the new Validate page, there is built-in functionality in the drop down search for adding tags that will sort based on a named field. I gave it a default name of "popularity" right now, and we would just need to add that property to the tag objects.
> 
> Here's an image from that UI: ![Screenshot from 2025-03-03 14-15-41](https://private-user-images.githubusercontent.com/6518824/418743127-16295471-9503-479e-baa1-7240e96992d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMzE2NzYsIm5iZiI6MTc0NDEzMTM3NiwicGF0aCI6Ii82NTE4ODI0LzQxODc0MzEyNy0xNjI5NTQ3MS05NTAzLTQ3OWUtYmFhMS03MjQwZTk2OTkyZDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDhUMTY1NjE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWRlZjliZDk1ZWIwOTMwNGFlODcxNmM2YzdmZTUwZDFjNzNlYWJjNjc4NGZmZDFlYzlmMjc0ZWM2ZDRmOTM3MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.RFCdBWqtixV5rlvNcZNW52fhylIi3oP_TCbmAtv3kRk)

@misaugstad thanks for the comments! I have addressed the sorting on the Explore page by making sure mutually exclusive tags are scored in a way that they are always next to one another: 
<img width="403" alt="image" src="https://github.com/user-attachments/assets/ab96fcd5-cceb-4c05-bd00-30fd69253165" />

I also changed the field from count to popularity so that new validate beta can have the correct sorting

bella-buchanan committed:
Merge branch 'develop' into 3578-order-tags-by-frequency-of-use

misaugstad commented:
Things are looking good on the Explore page, but on the NewValidateBeta, it still doesn't look sorted! Below are what I'm seeing, and then the output from the /label/tags API:
![Screenshot from 2025-04-08 10-53-20](https://github.com/user-attachments/assets/4c545fd6-8b70-4944-a038-11859983d339)

![Screenshot from 2025-04-08 10-53-52](https://github.com/user-attachments/assets/bd2e8797-2c94-4a06-beb4-ab98492c0c81)

bella-buchanan commented:
> Things are looking good on the Explore page, but on the NewValidateBeta, it still doesn't look sorted! Below are what I'm seeing, and then the output from the /label/tags API: ![Screenshot from 2025-04-08 10-53-20](https://private-user-images.githubusercontent.com/6518824/431509555-4c545fd6-8b70-4944-a038-11859983d339.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMzY4MDksIm5iZiI6MTc0NDEzNjUwOSwicGF0aCI6Ii82NTE4ODI0LzQzMTUwOTU1NS00YzU0NWZkNi04YjcwLTQ5NDQtYTAzOC0xMTg1OTk4M2QzMzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDhUMTgyMTQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWFkZTQ4ZWNkZTVmZmQ0N2M5YzRhN2NjZjZjMjcxYTI3YTYxY2M0NGUwN2ZhY2M1ZTg2ZTQxNGIxYmEzNmE5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hyjVt7Y7kn8oQHdPnXFVHbTWtwel7FmKO-okiYLiHX0)
> 
> ![Screenshot from 2025-04-08 10-53-52](https://private-user-images.githubusercontent.com/6518824/431509578-bd2e8797-2c94-4a06-beb4-ab98492c0c81.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMzY4MDksIm5iZiI6MTc0NDEzNjUwOSwicGF0aCI6Ii82NTE4ODI0LzQzMTUwOTU3OC1iZDJlODc5Ny0yYzk0LTRhMDYtYmViNC1hYjk4NDkyYzBjODEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDhUMTgyMTQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjhlZWU1OWQwZjc5M2FlMmQ4N2QzZDM3NDliNjFmYmY0OWEzNWI0OTlkMTEwY2Y2OTc2YzY3MzIyZjIxZmY3YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vIkvpNgc1uo0pXIVqWtkMLQHTjBj5FJfiRUgIm4q28w)

Thanks @misaugstad, I've been struggling to login as an admin so let me figure that out to close out the new validate issue you're seeing!

misaugstad commented:
@bella-buchanan check these docs for help with that!
https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Process-for-contributing-new-code#do-you-need-to-login-as-an-admin-user

bella-buchanan committed:
Merge branch 'develop' into 3578-order-tags-by-frequency-of-use

bella-buchanan committed:
actually do the sorting on the FE

bella-buchanan committed:
change name to count and fix sorting

bella-buchanan committed:
add count to tag object

bella-buchanan committed:
revert context menu

bella-buchanan committed:
Revert "revert context menu"

This reverts commit 65783a6f24007e5718584cc56f102d5eb310b933.

bella-buchanan committed:
include count in tag writes

bella-buchanan committed:
sort tags on FE in newValidateBeta

bella-buchanan committed:
common utility method to sort tags

bella-buchanan committed:
fix formatting dif

bella-buchanan committed:
type annotation

misaugstad commented on app/models/label/TagTable.scala:
```diff
@@ -8,15 +8,16 @@ import play.api.cache.Cache
 import scala.slick.lifted.{ForeignKeyQuery, Index}
 
-case class Tag(tagId: Int, labelTypeId: Int, tag: String, mutuallyExclusiveWith: Option[String])
+case class Tag(tagId: Int, labelTypeId: Int, tag: String, mutuallyExclusiveWith: Option[String], count: Int = 0)
```
It's a bit annoying, but let's keep the `Tag` class just containing the data in the table. We can just create a new case class if need be. For every table we have a case class with exactly that data, so I don't think it's worth it to break that convention.

misaugstad reviewed (changes requested):
A few things from testing:
1. You somehow removed the following lines from the end of the `escapeHTML` function in Utilities.js, which made it so that neither the Explore nor Validate page would load... Had to figure out what happened to be able to test in the first place.
    ```
            }
        });
    }
    ```
2. On Validate, it seems like maybe the tags start out sorted, but once you add or remove a tag, it removes the sorting..? Here's a before/after when adding a tag. It starts out sorted correctly, but sorting is totally different after adding a tag.
    ![Screenshot from 2025-04-17 15-26-22](https://github.com/user-attachments/assets/1f9bbb4b-472c-45dd-87a2-a27d737e3623)
    ![Screenshot from 2025-04-17 15-28-41](https://github.com/user-attachments/assets/5b1bb4cc-e34e-4e97-ba85-76df87b5666f)

bella-buchanan committed:
Merge branch 'develop' into 3578-order-tags-by-frequency-of-use

bella-buchanan committed:
Merge branch 'develop' into 3578-order-tags-by-frequency-of-use

bella-buchanan committed:
detailed tag case class

bella-buchanan committed:
fix validate tag display

jonfroehlich commented:
Checking in on this one, @bella-buchanan. Based on the last review thread (May 2025), there were two remaining issues after your last update:

1. **Validate page still unsorted** — sorting was working on the Explore page but the Validate page was still showing tags in the original order
2. **Mutually exclusive pairs not always adjacent** — e.g. "one button" / "two buttons" weren't next to each other in all cases

The branch also now has merge conflicts against `develop`. Are you able to pick this back up — address those two items and rebase? Happy to help if you run into anything.

misaugstad commented:
Would be easier to start from scratch at this point, closing now.
